arch/risc-v/src/common/riscv_misaligned.c

```diff
@@ -0,0 +1,418 @@
+        return false;
+    }
+
+  _alert("Misaligned compressed load/store: %x/%x\n", in, insn.insn);
+
+  regs[REG_EPC] += 2;
+
+  return true;
+#else
+  return false;
+#endif
+}
+
+/****************************************************************************
+ * Name: decode_insn_ctx
+ *
+ * Description:
+ *   Try to decode a normal encoding
+ *
+ ****************************************************************************/
+
+static bool decode_insn_ctx(uintptr_t *regs, rv_insn_ctx_t *ctx)
+{
+  uint32_t in;
+  rv_insn_t insn;
+  int32_t imm;
+
+  insn.insn = *(uint32_t *)regs[REG_EPC];
+
+  /* Adjust EPC */
+
+  regs[REG_EPC] += 4;
+
+  /* Get load/store instruction encoding */
+
+  in = insn.insn & INSN_MASK;
+
+  _alert("Misaligned load/store: %x/%x\n", in, insn.insn);
+
+  switch (in)
+    {
+      case INSN_LH:
+      case INSN_LHU:
+      case INSN_LW:
+      case INSN_LWU:
+      case INSN_LD:
+
+        /* Load a value from memory to register */
+
+        ctx->dest = (uint8_t *)&regs[insn.l.rd];
+        ctx->src = (uint8_t *)regs[insn.l.rs1] + insn.l.imm;
+
+        /* Zero the target register */
+
+        *(uintptr_t *)ctx->dest = 0;
+
+        /* Get data wide bit */
+
+        in &= (~0x4000);
+        in >>= 12;
+        ctx->len = 2 << (in - 1);
+
+        break;
+
+      case INSN_SH:
+      case INSN_SW:
+      case INSN_SD:
+
+        /* Fetch sign extended offset */
+
+        *(uint32_t *)&imm = insn.s.imm2 | (insn.s.imm1 << 5);
+
+        ctx->dest = (uint8_t *)regs[insn.s.rs1] + imm;
+        ctx->src = (uint8_t *)&regs[insn.s.rs2];
+
+        /* Get data wide bit */
+
+        in >>= 12;
+        ctx->len = 2 << (in - 1);
+
+        break;
+
+      case INSN_FLW:
+      case INSN_FLD:
+      case INSN_FSW:
+      case INSN_FSD:
+      default:
+        return false;
+    }
+
+  return true;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: riscv_misaligned
+ *
+ * Description:
+ *   This is software interrupt exception handler that handle load/store
+ *   address misaligned exception.
+ *
+ * Input Parameters:
+ *   regs - The current exception context
+ *
+ ****************************************************************************/
+
+int riscv_misaligned(int irq, void *context, void *arg)
+{
+  uintptr_t *regs = context;
+  rv_insn_ctx_t ctx;
+  bool ret;
+
+  /* Try to decode compressed instruction if it is */
+
+  ret = decode_insn_compressed(regs, &ctx);
+
+  /* Decode instruction context */
+
+  if (ret == false)
+    {
+      if (decode_insn_ctx(regs, &ctx) == false)
+        {
+          /* Decode failed, we can't handle a invalid instruction */
+
+          PANIC();
+        }
+    }
+
+  /* Byte copy */
+
+  while (ctx.len-- > 0)
+    {
+      *ctx.dest++ = *ctx.src++;
+    }
```

Comment:
probably can be replaced with `memcpy`. I mean that compiler will most probably optimize this.

Reply:
In fact, we can't do optimize here since it may introduce misaligned access then we will trigger a misaligned access exception in misaligned access exception handler.

Reply:
I didn't knew that memcpy can introduce missaligned access.

Reply:
This depends on the implementation of memcpy. I'm not sure target specific implementation like `libs/libc/machine/risc-v/rv32/arch_memcpy.S` is suitable or not.

For another issue, the maximum length is 8 (2/4/8), such a `optimized` implementation may not faster than byte copy.

Reply:
Anyway, let's leave optimization to compiler